engine/execution/ingestion/stop_control.go

```diff
@@ -3,318 +3,575 @@ package ingestion
+	// which is fine, but ideally we would stop right away
+
+	return sc
 }
 
-// IsPaused returns true is block execution has been paused
-func (s *StopControl) IsPaused() bool {
+// IsExecutionStopped returns true is block execution has been stopped
+func (s *StopControl) IsExecutionStopped() bool {
 	s.RLock()
 	defer s.RUnlock()
-	return s.state == StopControlPaused
+
+	return s.stopped
 }
 
-// SetStopHeight sets new stopHeight and crash mode, and return old values:
-//   - stopHeight
-//   - crash
-//
-// Returns error if the stopping process has already commenced, new values will be rejected.
-func (s *StopControl) SetStopHeight(
-	height uint64,
-	crash bool,
-) (uint64, bool, error) {
+// SetStop sets new stop parameters.
+// Returns error if the stopping process has already commenced, or if already stopped.
+func (s *StopControl) SetStop(
+	stop StopParameters,
+) error {
 	s.Lock()
 	defer s.Unlock()
 
-	oldHeight := s.stopHeight
-	oldCrash := s.crash
-
-	if s.state == StopControlCommenced {
-		return oldHeight,
-			oldCrash,
-			fmt.Errorf(
-				"cannot update stopHeight, "+
-					"stopping commenced for stopHeight %d with crash=%t",
-				oldHeight,
-				oldCrash,
-			)
+	stopBoundary := &stopBoundary{
+		StopParameters: stop,
 	}
 
-	if s.state == StopControlPaused {
-		return oldHeight,
-			oldCrash,
-			fmt.Errorf("cannot update stopHeight, already paused")
+	log := s.log.With().
+		Stringer("old_stop", s.stopBoundary).
+		Stringer("new_stop", stopBoundary).
+		Logger()
+
+	canChange, reason := s.canChangeStop(
+		stopBoundary.StopHeight,
+		false,
+	)
+	if !canChange {
+		err := fmt.Errorf(reason)
+
+		log.Warn().Err(err).Msg("cannot set stopHeight")
+		return err
 	}
 
-	// cannot set stopHeight to block which is already executing
-	// so the lowest possible stopHeight is highestExecutingHeight+1
-	if height <= s.highestExecutingHeight {
-		return oldHeight,
-			oldCrash,
-			fmt.Errorf(
-				"cannot update stopHeight, "+
-					"given stopHeight %d below or equal to highest executing height %d",
-				height,
-				s.highestExecutingHeight,
-			)
+	log.Info().Msg("stop set")
+	s.stopBoundary = stopBoundary
+
+	return nil
+}
+
+// canChangeStop verifies if the stop parameters can be changed
+// returns false and the reason if the parameters cannot be changed
+// if newHeight == math.MaxUint64 tha basically means that the stop is being removed
+func (s *StopControl) canChangeStop(
+	newHeight uint64,
+	fromVersionBeacon bool,
+) (
+	bool,
+	string,
+) {
+
+	if s.stopped {
+		return false, "cannot update stop parameters, already stopped"
 	}
 
-	s.log.Info().
-		Int8("previous_state", int8(s.state)).
-		Int8("new_state", int8(StopControlSet)).
-		Uint64("stopHeight", height).
-		Bool("crash", crash).
-		Uint64("old_height", oldHeight).
-		Bool("old_crash", oldCrash).
-		Msg("new stopHeight set")
+	if s.stopBoundary == nil {
+		// if there is no stop boundary set, we can set it to anything
```

Comment:
when there is no stop boundary set, can we initialize `stopBoundary` as if it's manually set to stop at MaxInt height without crashing. That way, we remove this special case where stopBoundary is nil